src/modules/cmdpal/ext/Microsoft.CmdPal.Ext.Bookmark/Services/BookmarkResolver.cs

```diff
@@ -0,0 +1,548 @@
+    {
+        // 1) Try URI parsing first (accepts custom schemes, e.g., shell:, ms-settings:)
+        //    File URIs must start with "file:" to avoid confusion with local paths - which are handled below, in more sophisticated ways -
+        //    as TryCreate would automatically add "file://" to bare paths like "C:\path\to\file.txt" which we don't want.
+        if (Uri.TryCreate(input, UriKind.Absolute, out var uri)
+            && !string.IsNullOrWhiteSpace(uri.Scheme)
+            && (uri.Scheme != Uri.UriSchemeFile || input.StartsWith("file:", StringComparison.OrdinalIgnoreCase))
+            && uri.Scheme != UriSchemeShell)
+        {
+            // http/https → Url; any other scheme → Protocol (mailto:, ms-settings:, slack://, etc.)
+            var isWeb = uri.Scheme == Uri.UriSchemeHttp || uri.Scheme == Uri.UriSchemeHttps;
+
+            result = new Classification(
+                isWeb ? CommandKind.WebUrl : CommandKind.Protocol,
+                input,
+                input,
+                string.Empty,
+                LaunchMethod.ShellExecute, // Shell picks the right handler
+                null,
+                isPlaceholder);
+
+            return true;
+        }
+
+        // 1a) We're a placeholder and start look like a protocol scheme (e.g. "myapp:{{placeholder}}")
+        if (isPlaceholder && !string.IsNullOrWhiteSpace(inputUntilFirstPlaceholder) && inputUntilFirstPlaceholder.Contains(':'))
+        {
+            var indexOfColon = inputUntilFirstPlaceholder.IndexOf(':');
+            var scheme = inputUntilFirstPlaceholder[..indexOfColon];
+
+            if (scheme.Length > 1 && scheme != Uri.UriSchemeFile && scheme != UriSchemeShell)
+            {
+                var isWeb = scheme.Equals(Uri.UriSchemeHttp, StringComparison.OrdinalIgnoreCase) || scheme.Equals(Uri.UriSchemeHttps, StringComparison.OrdinalIgnoreCase);
+
+                result = new Classification(
+                    isWeb ? CommandKind.WebUrl : CommandKind.Protocol,
+                    input,
+                    input,
+                    string.Empty,
+                    LaunchMethod.ShellExecute, // Shell picks the right handler
+                    null,
+                    isPlaceholder);
+                return true;
+            }
+        }
+
+        // 2) Existing file/dir or "longest plausible prefix"
+        // Try to grow head (only for unquoted original) to include spaces until a path exists.
+
+        // Find longest unquoted argument string
+        var (a, b) = CommandLineHelper.SplitLongestHeadBeforeQuotedArg(input);
+        if (a == string.Empty)
+        {
+            (a, b) = CommandLineHelper.SplitHeadAndArgs(input);
+        }
+
+        var (headPath, tailArgs) = ExpandToBestExistingPath(a, b, isPlaceholder, placeholderParser);
+        if (headPath is not null)
+        {
+            var args = tailArgs ?? string.Empty;
+
+            if (Directory.Exists(headPath))
```

Comment:
Don't we know that the `headPath` definitely does exist (either as a dir or file) here?
* `ExpandToBestExistingPath` calls down to
  * `GreedyFind` which calls to
    * `ExpandPathToPhysicalFile` which only returns true if the path exists
  * meaning that `GreedyFind` will only ever return a non-`null` `Head` if the `Head` exists
* so the non-null `headPath` result from `ExpandToBestExistingPath` does exist

(this is picking nits over essentially a single `File.Exists` call isn't it)